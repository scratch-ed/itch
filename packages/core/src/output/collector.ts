import { Diff, Group, Judgement, NestedGroup, Test, TestGroup } from './full-schema';
import {
  AppendDiff,
  AppendMessage,
  CloseGroup,
  CloseJudgement,
  CloseTest,
  EscalateStatus,
  StartGroup,
  StartJudgement,
  StartTest,
  Update,
} from './schema';

export class OutputCollector {
  private levelStack: Array<
    Array<Test | Group | AppendMessage | AppendDiff | EscalateStatus>
  > = [];

  private startCloseStack: Array<StartJudgement | StartGroup | StartTest> = [];

  public judgement?: Judgement;

  public handle(update: Update) {
    switch (update.command) {
      case 'start-judgement':
      case 'start-group':
      case 'start-test':
        this.startCloseStack.push(update);
        this.levelStack.push([]);
        break;
      case 'append-message':
      case 'append-diff':
      case 'escalate-status':
        this.currentLevel().push(update);
        break;
      case 'close-test': {
        const test = this.handleTestLevel(update);
        this.currentLevel().push(test);
        break;
      }
      case 'close-group': {
        const group = this.handleGroupLevel(update);
        this.currentLevel().push(group);
        break;
      }
      case 'close-judgement': {
        this.judgement = this.handleJudgementLevel(update);
        break;
      }
    }
  }

  private currentLevel() {
    return this.levelStack[this.levelStack.length - 1];
  }

  private handleTestLevel(finalUpdate: CloseTest): Test {
    const startUpdate = this.startCloseStack.pop()!;
    if (startUpdate.command !== 'start-test') {
      throw new Error(`Illegal update ${startUpdate} found in update stack.`);
    }

    const messages = [];
    let diff;
    let escalatedStatus;

    const levelObjects = this.levelStack.pop() as Array<
      AppendDiff | AppendMessage | EscalateStatus
    >;
    for (const levelObject of levelObjects!) {
      switch (levelObject.command) {
        case 'append-diff': {
          diff = new Diff(levelObject.expected, levelObject.actual);
          break;
        }
        case 'append-message': {
          messages.push(levelObject.message);
          break;
        }
        case 'escalate-status': {
          escalatedStatus = levelObject;
          break;
        }
        default: {
          throw new Error(`Illegal update ${levelObject} found in test level.`);
        }
      }
    }

    if (escalatedStatus) {
      this.currentLevel().push(escalatedStatus);
    }

    return new Test(
      startUpdate.name,
      escalatedStatus?.status || finalUpdate.status,
      finalUpdate.feedback,
      messages,
      diff,
    );
  }

  private handleGroupLevel(finalUpdate: CloseGroup): Group {
    const startUpdate = this.startCloseStack.pop()!;
    if (startUpdate.command !== 'start-group') {
      throw new Error(`Illegal update ${startUpdate} found in update stack.`);
    }

    const messages: string[] = [];
    const groups: Group[] = [];
    const tests: Test[] = [];
    let escalatedStatus;

    const levelObjects = this.levelStack.pop() as Array<
      AppendMessage | Test | Group | EscalateStatus
    >;
    for (const levelObject of levelObjects!) {
      if (levelObject instanceof Test) {
        tests.push(levelObject);
      } else if (levelObject instanceof NestedGroup || levelObject instanceof TestGroup) {
        groups.push(levelObject);
      } else if (levelObject.command === 'append-message') {
        messages.push(levelObject.message);
      } else if (levelObject.command === 'escalate-status') {
        escalatedStatus = levelObject;
      } else {
        throw new Error(`Illegal update ${levelObject} found in group level.`);
      }
    }

    if (escalatedStatus) {
      this.currentLevel().push(escalatedStatus);
    }

    if (tests.length > 0 && groups.length > 0) {
      throw new Error(`Cannot mix groups and tests as children of a group.`);
    }
    if (tests.length > 0) {
      return new TestGroup(
        tests,
        startUpdate.name,
        startUpdate.visibility,
        startUpdate.sprite,
        messages,
        escalatedStatus?.status,
        finalUpdate.summary,
      );
    } else {
      return new NestedGroup(
        groups,
        startUpdate.name,
        startUpdate.visibility,
        startUpdate.sprite,
        messages,
        escalatedStatus?.status,
        finalUpdate.summary,
      );
    }
  }

  private handleJudgementLevel(_update: CloseJudgement): Judgement {
    const startUpdate = this.startCloseStack.pop()!;
    if (startUpdate.command !== 'start-judgement') {
      throw new Error(`Illegal update ${startUpdate} found in update stack.`);
    }

    const messages: string[] = [];
    const groups: Group[] = [];
    let escalatedStatus;

    const levelObjects = this.levelStack.pop() as Array<
      AppendMessage | Group | EscalateStatus
    >;
    for (const levelObject of levelObjects!) {
      if (levelObject instanceof NestedGroup || levelObject instanceof TestGroup) {
        groups.push(levelObject);
      } else if (levelObject.command === 'append-message') {
        messages.push(levelObject.message);
      } else if (levelObject.command === 'escalate-status') {
        escalatedStatus = levelObject;
      } else {
        throw new Error(`Illegal update ${levelObject} found in judgement level.`);
      }
    }

    return new Judgement(groups, messages, escalatedStatus?.status);
  }
}
