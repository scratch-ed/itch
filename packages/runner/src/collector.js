/**
 * @fileOverview A basic debug view for the output of the judge.
 */
class Processor {

  constructor(element) {
    this.stack = [];
    this.element = element;
  }

  process(message) {

    /** @type {string} */
    const command = message.command;
    
    if (command === 'start-judgement') {
      this.start = new Date();
    }
    
    if (command === 'close-judgement') {
      const title = document.getElementById("out");
      if (title) {
        title.innerHTML = `Output (took ${new Date() - this.start}ms)`;
      }
    }

    if (command.startsWith('start')) {
      this.stack.push(message);
    }

    if (command === 'start-tab') {
      this.output(`<h2>${message.title || 'Tabblad'}</h2>`);
    } else if (command === 'start-context') {
      this.output(`<h3>${message.description || 'Context'}</h3>`);
    } else if (command === 'start-testcase') {
      this.case = message.description;
      this.caseTests = [];
    } else if (command === "start-test") {
      this.currentTest = message;
    } else if (command === 'close-test') {
      const expected = this.currentTest;
      const actual = message.generated;
      const status = message.status.enum;
      this.caseTests.push({
        expected: expected,
        actual: actual,
        status: status
      });
    } else if (command === "close-testcase") {
      const content = this.caseTests.map(test => {
        return `${test.status === 'correct' ? '✅' : '❌'} Expected ${test.expected.expected}, got: ${test.actual}`;
      });
      const merged = content.join("\n");
      const allCorrect = this.caseTests.every(test => test.status === 'correct');
      this.output(`<span title="${this.case}\n${merged}">${allCorrect ? '✅' : '❌'}</span>`);
    } else if (command.startsWith('close')) {
      this.stack.pop();
    } else if (command === 'append-message') {
      this.output(`<br>Message: ${message.message}`);
    }
  }

  output(res) {
    this.element.innerHTML += this.prepareHtml(res);
  }

  info(res) {
    this.element.innerHTML += this.prepareHtml(res);
  }

  error(res) {
    this.element.innerHTML += this.prepareHtml(res);
  }

  /** @param {string} html */
  prepareHtml(html) {
    return html;
  }
}

// eslint-disable-next-line no-unused-vars
function init(element) {
  const old = window.handleOut;
  const processor = new Processor(element);
  window.handleOut = (result) => {
    processor.process(result);
    old(result);
  };
}