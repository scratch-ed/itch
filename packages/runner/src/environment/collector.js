/**
 * @fileOverview A basic debug view for the output of the judge.
 */
class Processor {
  constructor(element) {
    this.stack = [];
    this.element = element;
    this.groupStack = [];
  }

  process(message) {
    /** @type {string} */
    const command = message.command;

    if (command === 'start-judgement') {
      this.start = new Date();
    }

    if (command === 'close-judgement') {
      const title = document.getElementById('out');
      if (title) {
        title.innerHTML = `Output (took ${new Date() - this.start}ms)`;
      }
    }

    if (command.startsWith('start')) {
      this.stack.push(message);
    }

    if (command === 'start-group') {
      // Start a new group.
      const details = document.createElement('details');
      details.open = message.visibility === 'show';
      const summary = document.createElement('summary');
      summary.innerHTML = `<strong>${message.name}</strong>`;
      details.appendChild(summary);
      this.element.appendChild(details);
      this.groupStack.push(this.element);
      this.element = details;
    }

    if (command === 'close-group') {
      if (message.description) {
        const summary = this.element.querySelector('summary');
        summary.innerHTML += '<br>';
        summary.innerHTML += message.description;
      }
      this.element = this.groupStack.pop();
    }

    if (command === 'start-test') {
      this.currentTest = message;
    }

    if (command === 'close-test') {
      this.element.innerHTML += `<span title='${this.currentTest.name}'>${
        message.status === 'correct' ? '✅' : '❌'
      } ${message.feedback || this.currentTest.name}</span><br>`;
    }

    if (command.startsWith('close')) {
      this.stack.pop();
    } else if (command === 'append-message') {
      this.element.innerHTML += `Message: ${message.message}`;
    } else if (command === 'escalate-status') {
      this.element.innerHTML += `<br><strong>ESCALATION</strong>: status is now ${message.status}`;
    }
  }
}

// eslint-disable-next-line no-unused-vars
function init(element) {
  const processor = new Processor(element);
  window.handleOut = (result) => {
    processor.process(result);
    console.debug(result);
  };
}
