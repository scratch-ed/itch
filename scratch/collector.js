class Processor {

  constructor(element) {
    this.stack = [];
    this.element = element;
  }

  process(message) {

    /** @type {string} */
    const command = message.command;

    if (command.startsWith('start')) {
      this.stack.push(message);
    }

    if (command === 'start-context') {
      this.output(`<h2>${message.description || 'Context'}</h2>`);
    } else if (command === 'start-testcase') {
      this.case = message.description;
      this.caseTests = [];
    } else if (command === "start-test") {
      this.currentTest = message;
    } else if (command === 'close-test') {
      const expected = this.currentTest;
      const actual = message.generated;
      const status = message.status.enum;
      this.caseTests.push({
        expected: expected,
        actual: actual,
        status: status
      });
    } else if (command === "close-testcase") {
      const content = this.caseTests.map(test => {
        return `${test.status === 'correct' ? '✅' : '❌'} Expected ${test.expected.expected}, got: ${test.actual}\n`;
      });
      const allCorrect = this.caseTests.every(test => test.status === 'correct');
      this.output(`<span title="${this.case}\n${content}">${allCorrect ? '✅' : '❌'}</span>`);
    } else if (command.startsWith('close')) {
      this.stack.pop();
    } else if (command === 'append-message') {
      this.output(`Message: ${message.message}`);
    }
  }

  output(res) {
    this.element.innerHTML += this.prepareHtml(res);
  }

  info(res) {
    this.element.innerHTML += this.prepareHtml(res);
  }

  error(res) {
    this.element.innerHTML += this.prepareHtml(res);
  }

  /** @param {string} html */
  prepareHtml(html) {
    return html;
  }
}

function init(element) {
  const old = window.handleOut;
  const processor = new Processor(element);
  window.handleOut = (result) => {
    processor.process(result);
    old(result);
  };
}