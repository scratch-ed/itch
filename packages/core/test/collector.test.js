import { OutputCollector } from '../src/output/collector.ts';
import source from './partial.json';

test('Collector works correctly', () => {
  const collector = new OutputCollector();
  source.forEach((update) => collector.handle(update));
  expect(collector.judgement).toMatchSnapshot();
});
