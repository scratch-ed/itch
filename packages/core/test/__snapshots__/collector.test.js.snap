// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collector works correctly 1`] = `
Judgement {
  "children": [
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "false",
            "expected": "false",
          },
          "feedback": "Top! Je hebt niets veranderd aan de sprite Scherm.",
          "messages": [],
          "name": "Scherm",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "false",
            "expected": "false",
          },
          "feedback": "Top! Je hebt niets veranderd aan de blokjes van sprite Scherm.",
          "messages": [],
          "name": "Scherm",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "false",
            "expected": "false",
          },
          "feedback": "Top! Je hebt niets veranderd aan de sprite Boodschap 1.",
          "messages": [],
          "name": "Boodschap 1",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "false",
            "expected": "false",
          },
          "feedback": "Top! Je hebt niets veranderd aan de blokjes van sprite Boodschap 1.",
          "messages": [],
          "name": "Boodschap 1",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "false",
            "expected": "false",
          },
          "feedback": "Top! Je hebt niets veranderd aan het speelveld.",
          "messages": [],
          "name": "Speelveld",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "false",
            "expected": "false",
          },
          "feedback": "Top! Je hebt niets veranderd aan de blokjes van het speelveld.",
          "messages": [],
          "name": "Speelveld",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Goed zo! Je hebt geen losse blokjes laten rondslingeren.",
          "messages": [],
          "name": "Planeet",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Je hebt niets veranderd aan de voorgeprogrammeerde blokjes van de sprite Planeet.",
          "messages": [],
          "name": "Planeet",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Goed zo! Je hebt geen losse blokjes laten rondslingeren.",
          "messages": [],
          "name": "Vijand 1",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Je hebt niets veranderd aan de voorgeprogrammeerde blokjes van de sprite Vijand 1.",
          "messages": [],
          "name": "Vijand 1",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Goed zo! Je hebt geen losse blokjes laten rondslingeren.",
          "messages": [],
          "name": "Vijand 2",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Je hebt niets veranderd aan de voorgeprogrammeerde blokjes van de sprite Vijand 2.",
          "messages": [],
          "name": "Vijand 2",
          "status": "runtime error",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Goed zo! Je hebt geen losse blokjes laten rondslingeren.",
          "messages": [],
          "name": "Voedsel 1",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Je hebt niets veranderd aan de voorgeprogrammeerde blokjes van de sprite Voedsel 1.",
          "messages": [],
          "name": "Voedsel 1",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Goed zo! Je hebt geen losse blokjes laten rondslingeren.",
          "messages": [],
          "name": "Voedsel 2",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Je hebt niets veranderd aan de voorgeprogrammeerde blokjes van de sprite Voedsel 2.",
          "messages": [],
          "name": "Voedsel 2",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Goed zo! Je hebt geen losse blokjes laten rondslingeren.",
          "messages": [],
          "name": "Voedsel 3",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Je hebt niets veranderd aan de voorgeprogrammeerde blokjes van de sprite Voedsel 3.",
          "messages": [],
          "name": "Voedsel 3",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Goed zo! Je hebt geen losse blokjes laten rondslingeren.",
          "messages": [],
          "name": "Voedsel 4",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Je hebt niets veranderd aan de voorgeprogrammeerde blokjes van de sprite Voedsel 4.",
          "messages": [],
          "name": "Voedsel 4",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Controle op bestaande code",
      "sprite": undefined,
      "status": "runtime error",
      "summary": undefined,
      "visibility": "show",
    },
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "control_forever",
            "expected": "control_forever",
          },
          "feedback": "De planeet kan oneindig lang bewegen.",
          "messages": [],
          "name": "Juiste lus",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "De planeet richt zich naar de muis.",
          "messages": [],
          "name": "Richt naar de muis",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "De planeet zet 5 stappen.",
          "messages": [],
          "name": "Beweegt",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "De planeet verandert zijn grootte.",
          "messages": [],
          "name": "Planeet verandert grootte",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "De planeet verandert zijn snelheid.",
          "messages": [],
          "name": "Planeet verandert snelheid",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Test voor Planeet.",
      "sprite": undefined,
      "status": undefined,
      "summary": undefined,
      "visibility": "show",
    },
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "control_forever",
            "expected": "control_forever",
          },
          "feedback": "De planeet kan oneindig lang bewegen.",
          "messages": [],
          "name": "Juiste lus",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Vijand 1 beweegt zich binnen de 5 seconden naar een willekeurige positie.",
          "messages": [],
          "name": "Richt naar de muis",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Test voor Vijand 1.",
      "sprite": undefined,
      "status": undefined,
      "summary": undefined,
      "visibility": "show",
    },
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "control_forever",
            "expected": "control_forever",
          },
          "feedback": "De planeet kan oneindig lang bewegen.",
          "messages": [],
          "name": "Juiste lus",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Vijand 2 beweegt zich binnen de 5 seconden naar een willekeurige positie.",
          "messages": [],
          "name": "Richt naar de muis",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Test voor Vijand 2.",
      "sprite": undefined,
      "status": undefined,
      "summary": undefined,
      "visibility": "show",
    },
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "control_forever",
            "expected": "control_forever",
          },
          "feedback": "De planeet kan oneindig lang bewegen.",
          "messages": [],
          "name": "Juiste lus",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_hide",
            "expected": "looks_hide",
          },
          "feedback": "Voedsel 1 begint onzichtbaar.",
          "messages": [],
          "name": "Verdwijnt",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 1 wacht één seconde.",
          "messages": [],
          "name": "Wacht",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 1 gaat naar een willekeurige positie.",
          "messages": [],
          "name": "Willekeurige positie",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_show",
            "expected": "looks_show",
          },
          "feedback": ":AIO_TP_VerschijntJuist",
          "messages": [],
          "name": "Verschijn",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 1 beweegt in een willekeurige richting tot het de planeet raakt.",
          "messages": [],
          "name": "Beweeg tot raak",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Test voor Voedsel 1",
      "sprite": undefined,
      "status": undefined,
      "summary": undefined,
      "visibility": "show",
    },
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "control_forever",
            "expected": "control_forever",
          },
          "feedback": "De planeet kan oneindig lang bewegen.",
          "messages": [],
          "name": "Juiste lus",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_hide",
            "expected": "looks_hide",
          },
          "feedback": "Voedsel 2 begint onzichtbaar.",
          "messages": [],
          "name": "Verdwijnt",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 2 wacht één seconde.",
          "messages": [],
          "name": "Wacht",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 2 gaat naar een willekeurige positie.",
          "messages": [],
          "name": "Willekeurige positie",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_show",
            "expected": "looks_show",
          },
          "feedback": ":AIO_TP_VerschijntJuist",
          "messages": [],
          "name": "Verschijn",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 2 beweegt in een willekeurige richting tot het de planeet raakt.",
          "messages": [],
          "name": "Beweeg tot raak",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Test voor Voedsel 2",
      "sprite": undefined,
      "status": undefined,
      "summary": undefined,
      "visibility": "show",
    },
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "control_forever",
            "expected": "control_forever",
          },
          "feedback": "De planeet kan oneindig lang bewegen.",
          "messages": [],
          "name": "Juiste lus",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_hide",
            "expected": "looks_hide",
          },
          "feedback": "Voedsel 3 begint onzichtbaar.",
          "messages": [],
          "name": "Verdwijnt",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 3 wacht één seconde.",
          "messages": [],
          "name": "Wacht",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 3 gaat naar een willekeurige positie.",
          "messages": [],
          "name": "Willekeurige positie",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_show",
            "expected": "looks_show",
          },
          "feedback": ":AIO_TP_VerschijntJuist",
          "messages": [],
          "name": "Verschijn",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 3 beweegt in een willekeurige richting tot het de planeet raakt.",
          "messages": [],
          "name": "Beweeg tot raak",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Test voor Voedsel 3",
      "sprite": undefined,
      "status": undefined,
      "summary": undefined,
      "visibility": "show",
    },
    Group {
      "children": [
        Test {
          "diff": Diff {
            "actual": "control_forever",
            "expected": "control_forever",
          },
          "feedback": "De planeet kan oneindig lang bewegen.",
          "messages": [],
          "name": "Juiste lus",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_hide",
            "expected": "looks_hide",
          },
          "feedback": "Voedsel 4 begint onzichtbaar.",
          "messages": [],
          "name": "Verdwijnt",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 4 wacht één seconde.",
          "messages": [],
          "name": "Wacht",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 4 gaat naar een willekeurige positie.",
          "messages": [],
          "name": "Willekeurige positie",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "looks_show",
            "expected": "looks_show",
          },
          "feedback": ":AIO_TP_VerschijntJuist",
          "messages": [],
          "name": "Verschijn",
          "status": "correct",
        },
        Test {
          "diff": Diff {
            "actual": "true",
            "expected": "true",
          },
          "feedback": "Voedsel 4 beweegt in een willekeurige richting tot het de planeet raakt.",
          "messages": [],
          "name": "Beweeg tot raak",
          "status": "correct",
        },
      ],
      "messages": [],
      "name": "Test voor Voedsel 4",
      "sprite": undefined,
      "status": undefined,
      "summary": undefined,
      "visibility": "show",
    },
  ],
  "messages": [],
  "meta": Meta {
    "correctTests": 53,
    "totalTests": 53,
  },
  "status": "runtime error",
  "version": 2,
}
`;
